### PR TITLE
[libc++] Remove <queue> and <stack> includes from <format>

### PR DESCRIPTION
**Pull request by philnik777**
This reduces the include time of <format> from 691ms to 556ms.

### CODE REVIEW

**Review by ldionne — COMMENTED** [review]
Do we plan on doing a lot of these removals? If so, we should discuss whether to stabilize the transitive includes in C++20, since C++20 has started getting wide adoption and this *is* going to cause some breakage.

File: libcxx/include/__format/container_adaptor.h [hunk 1]
```diff
@@ -18,11 +18,11 @@
 #include <__format/concepts.h>
 #include <__format/formatter.h>
 #include <__format/range_default_formatter.h>
+#include <__fwd/queue.h>
+#include <__fwd/stack.h>
```
Comment (frederick-vs-ja) [inline libcxx/include/__format/container_adaptor.h:22]: Why do we need place these concrete `formatter` specializations in a concentrated header?
IIRC these specializations should be placed in `<stack>` and `<queue>`, see cplusplus/draft#6635. (The change was not so editorial to me, but Daniel told me it was.)

File: libcxx/include/format [hunk 2]
```diff
@@ -223,6 +223,8 @@ namespace std {
 
 #if !defined(_LIBCPP_REMOVE_TRANSITIVE_INCLUDES) && _LIBCPP_STD_VER <= 20
 #  include <locale>
+#  include <queue>
+#  include <stack>
```
Comment (frederick-vs-ja) [inline libcxx/include/format:227]: I don't know why the inclusions are wanted here... I think `<format>` should be unrelated to `<stack>` and `<queue>` when `_LIBCPP_STD_VER <= 20`. The inclusion of `<stack>` and `<queue>` was introduced in https://github.com/llvm/llvm-project/commit/04d4f4b3d4e4fd608a3bc3fe387006435f04b61d (https://reviews.llvm.org/D141290) which just implemented some parts of [P2286R8](https://wg21.link/p2286r8) (a C++23 paper).

Comment (philnik777) [reply libcxx/include/format:227]: https://libcxx.llvm.org/DesignDocs/HeaderRemovalPolicy.html

File: libcxx/include/__format/container_adaptor.h (hunk 1 above)
Comment (philnik777) [reply libcxx/include/__format/container_adaptor.h:22]: IDK, but this seems not really related to the change made here. FWIW the change doesn't look editorial to me either.

**Comment by philnik777** [comment]
> Do we plan on doing a lot of these removals? If so, we should discuss whether to stabilize the transitive includes in C++20, since C++20 has started getting wide adoption and this _is_ going to cause some breakage.

The transitive includes in C++20 are stable. This only changes the transitive includes in C++23 and 26.

**Comment by mordante** [comment]
> Do we plan on doing a lot of these removals? If so, we should discuss whether to stabilize the transitive includes in C++20, since C++20 has started getting wide adoption and this _is_ going to cause some breakage.

@ldionne Transitive includes in C++20 have been "stable" (as in no removals) for several year. The last monthly meeting I mentioned C++23 but there was no real consensus to keep C++23 stable.

File: libcxx/include/iosfwd [hunk 3]
```diff
@@ -162,10 +163,6 @@ using wosyncstream = basic_osyncstream<wchar_t>;
 
 #endif // _LIBCPP_STD_VER >= 20 && !defined(_LIBCPP_HAS_NO_EXPERIMENTAL_SYNCSTREAM)
 
-// Include other forward declarations here
-template <class _Tp, class _Alloc = allocator<_Tp> >
-class _LIBCPP_TEMPLATE_VIS vector;
```
Comment (mordante) [inline libcxx/include/iosfwd:167]: Why if `__fwd/vector` not included?

File: libcxx/include/module.modulemap [hunk 4]
```diff
@@ -1251,6 +1251,8 @@ module std_private_debug_utils_strict_weak_ordering_check [system] {
   export std_private_type_traits_is_constant_evaluated
 }
 
+module std_private_deque_fwd [system] { header "__fwd/deque.h" }
```
Comment (mordante) [inline libcxx/include/module.modulemap:1254]: Please align `[system]` with `[system]` on line 1256.

File: libcxx/include/module.modulemap [hunk 5]
```diff
@@ -1596,6 +1599,8 @@ module std_private_numeric_transform_exclusive_scan [system] { header "__numeric
 module std_private_numeric_transform_inclusive_scan [system] { header "__numeric/transform_inclusive_scan.h" }
 module std_private_numeric_transform_reduce         [system] { header "__numeric/transform_reduce.h" }
 
+module std_private_queue_fwd [system] { header "__fwd/queue.h" }
```
Comment (mordante) [inline libcxx/include/module.modulemap:1602]: There too.

File: libcxx/include/module.modulemap [hunk 6]
```diff
@@ -1733,6 +1738,8 @@ module std_private_ranges_zip_view                   [system] { header "__ranges
 
 module std_private_span_span_fwd [system] { header "__fwd/span.h" }
 
+module std_private_stack_fwd [system] { header "__fwd/stack.h" }
+
```
Comment (mordante) [inline libcxx/include/module.modulemap:1742]: And here and it would be nice to do span too as a drive-by.

**Review by mordante — COMMENTED** [review]
Thanks for working on this. I actually like the patch for 2 reasons
- centralizing forward declarations in one place
- reducing compilation time
Please mention the first in the commit message too; it does not only improve `format`.

In general quite happy, I'd like to have a quick look after the review comments have been addressed.

File: libcxx/include/module.modulemap (hunk 4 above)
Comment (philnik777) [reply libcxx/include/module.modulemap:1254]: We align them per subdirectory in the modulemap AFAICT. Why should we change that here?

**Review by ldionne — APPROVED** [review]
Apologies, I thought this modified transitive includes in C++20.

File: libcxx/include/iosfwd (hunk 3 above)
Comment (philnik777) [reply libcxx/include/iosfwd:167]: Because it's not part of `<iosfwd>`.

**Comment by nico** [comment]
Compile time when including standard library headers is a major pain point, so thank you for working on this!

This has the effect of making `<string>` no longer include `<vector>`. That's great -- but it also requires some amount of cleanup. For our codebase it wasn't terrible, but a bit annoying. We use C++20, but we don't use `<format>` anywhere directly, and some people maintain the linux build building with libstdc++. Still, we had to clean up a bunch of files on a bunch of platforms where we exclusively use libc++ (...even though we don't use <format> ourselves).

I imagine it might be somewhat annoying for larger codebases. There's probably not too much we can do about this though.

The only thing I can think of is to be mindful of not adding unnecessary includes in the first place when adding new headers or when adding features. How do you find which includes you can remove? Maybe it's possible to do something similar on reviews when new code gets added?

To be clear, no action needed and I'm very appreciative of this change and changes like it. Just thinking out loud some on how to maybe make this a bit easier on libc++'s users.

**Comment by philnik777** [comment]
@nico In chromium you actually make the conscious decision to opt-in to removing transitive includes with `_LIBCPP_REMOVE_TRANSITIVE_INCLUDES`. By default these includes aren't removed currently.

I basically look at the compile times and when I think "why TF is this included" I try to remove it. Not exactly an elegant process, but the best I came up with until now.

**Comment by nico** [comment]
> @nico In chromium you actually make the conscious decision to opt-in to removing transitive includes with `_LIBCPP_REMOVE_TRANSITIVE_INCLUDES`. By default these includes aren't removed currently.

Ah, cool.

I'll say that we added `_LIBCPP_REMOVE_TRANSITIVE_INCLUDES` 1.5 years ago after completing some cleanup back then. Projects committed to keeping libc++ up-to-date with includes kind of had to define it back then to prevent regressions. It wasn't clear to me that this is covering new include cleanups in perpetuity. For bigger include cleanups, it would be useful to have one of these defines per cleanup :P (…but there haven't been any large cleanups needed since we added `_LIBCPP_REMOVE_TRANSITIVE_INCLUDES` 1.5 years ago, so it's not an urgent problem or anything).
